mh/pat.py

```diff
@@ -913,6 +913,33 @@ def recvReqLayerEnd(self, packet_id, data, seq):
         JP: レイヤ終了要求
         TR: Layer end request
         """
+        if self.session.layer > 1:
+            # City path
+            new_leader = self.session.try_transfer_city_leadership()
+            if new_leader is not None:
+                self.sendNtcLayerHost(new_leader, seq)
+
+            ntc_data = pati.lp2_string(self.session.capcom_id)
+            self.server.layer_broadcast(self.session, PatID4.NtcLayerOut,
+                                        ntc_data, seq)
+            self.session.leave_city()
```

Comment:
IMHO, this whole `session.leave_` fallthrough should be done in the `session.layer_end` method.

Reply:
I'll implement it this way, but be aware that this will require some form of packet sending to be done in the `session` class. I can minimize this by passing them in as callback functions, but some form of sending logic will still need to be in `session` (such as the `seq` value now needing to be passed into `session.layer_end`)

Reply:
I changed it in latest commit using callbacks, minimizing the amount of `session` logic in `fmp_server`. But let me know if you want me to switch it to what you've put here

Reply:
The callback idea is interesting but I think it shouldn't be used here. I'll keep that idea in mind as it will indeed be useful for implementing custom broadcast logic.

Reply:
Callback was removed in recent commit and changed to recommended logic.